CMake is not generating unbound/config.h correctly with openssl 1.1

This example uses debian stretch, aarch64, openssl 1.1.0f:

https://build.getmonero.org/builders/monero-static-debian-armv8/builds/1737/steps/compile/logs/stdio

/home/buildbot/slave/monero-static-debian-armv8/build/external/unbound/validator/val_secalgo.c:637:34: error: invalid application of 'sizeof' to an incomplete type 
                 'EVP_MD_CTX' (aka 'struct evp_md_ctx_st') ctx = (EVP_MD_CTX*)malloc(sizeof(*ctx));

On IRC hyc confirms the same issue for him with openssl 1.1

Manually editing unbound/config.h to #define HAVE_EVP_MD_CTX_NEW works.

This issue is to solicit help figuring out why configure_checks.cmake doesn't find EVP_MD_CTX_new

Determining if the function EVP_MD_CTX_new exists failed with the following output:
Change Dir: /home/buildbot/slave/monero-static-debian-armv8/build/build/release/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/make" "cmTC_34550/fast"
/usr/bin/make -f CMakeFiles/cmTC_34550.dir/build.make CMakeFiles/cmTC_34550.dir/build
make[1]: Entering directory '/home/buildbot/slave/monero-static-debian-armv8/build/build/release/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_34550.dir/CheckFunctionExists.c.o
/usr/bin/cc    -DCHECK_FUNCTION_EXISTS=EVP_MD_CTX_new   -o CMakeFiles/cmTC_34550.dir/CheckFunctionExists.c.o   -c /usr/share/cmake-3.7/Modules/CheckFunctionExists.c
Linking C executable cmTC_34550
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_34550.dir/link.txt --verbose=1
/usr/bin/cc   -DCHECK_FUNCTION_EXISTS=EVP_MD_CTX_new    CMakeFiles/cmTC_34550.dir/CheckFunctionExists.c.o  -o cmTC_34550 -rdynamic /usr/lib/aarch64-linux-gnu/libssl.a /usr/lib/aarch64-linux-gnu/libcrypto.a -ldl 
/usr/lib/aarch64-linux-gnu/libcrypto.a(threads_pthread.o): In function `CRYPTO_THREAD_lock_new':
(.text+0x30): undefined reference to `pthread_rwlock_init'
/usr/lib/aarch64-linux-gnu/libcrypto.a(threads_pthread.o): In function `CRYPTO_THREAD_read_lock':
(.text+0x68): undefined reference to `pthread_rwlock_rdlock'
/usr/lib/aarch64-linux-gnu/libcrypto.a(threads_pthread.o): In function `CRYPTO_THREAD_write_lock':
(.text+0x88): undefined reference to `pthread_rwlock_wrlock'
/usr/lib/aarch64-linux-gnu/libcrypto.a(threads_pthread.o): In function `CRYPTO_THREAD_unlock':
(.text+0xa8): undefined reference to `pthread_rwlock_unlock'
/usr/lib/aarch64-linux-gnu/libcrypto.a(threads_pthread.o): In function `CRYPTO_THREAD_lock_free':
(.text+0xd4): undefined reference to `pthread_rwlock_destroy'
/usr/lib/aarch64-linux-gnu/libcrypto.a(threads_pthread.o): In function `CRYPTO_THREAD_run_once':
(.text+0x108): undefined reference to `pthread_once'
/usr/lib/aarch64-linux-gnu/libcrypto.a(threads_pthread.o): In function `CRYPTO_THREAD_init_local':
(.text+0x128): undefined reference to `pthread_key_create'
/usr/lib/aarch64-linux-gnu/libcrypto.a(threads_pthread.o): In function `CRYPTO_THREAD_get_local':
(.text+0x144): undefined reference to `pthread_getspecific'
/usr/lib/aarch64-linux-gnu/libcrypto.a(threads_pthread.o): In function `CRYPTO_THREAD_set_local':
(.text+0x154): undefined reference to `pthread_setspecific'
/usr/lib/aarch64-linux-gnu/libcrypto.a(threads_pthread.o): In function `CRYPTO_THREAD_cleanup_local':
(.text+0x174): undefined reference to `pthread_key_delete'
collect2: error: ld returned 1 exit status
CMakeFiles/cmTC_34550.dir/build.make:99: recipe for target 'cmTC_34550' failed
make[1]: *** [cmTC_34550] Error 1
make[1]: Leaving directory '/home/buildbot/slave/monero-static-debian-armv8/build/build/release/CMakeFiles/CMakeTmp'
Makefile:126: recipe for target 'cmTC_34550/fast' failed
make: *** [cmTC_34550/fast] Error 2

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -319,6 +319,12 @@ else()
   message(STATUS "Stack trace on exception disabled")
 endif()
 
+if (UNIX AND NOT APPLE)
+  # Note that at the time of this writing the -Wstrict-prototypes flag added below will make this fail
+  set(THREADS_PREFER_PTHREAD_FLAG ON)
+  find_package(Threads)
+endif()
+
 # Handle OpenSSL, used for sha256sum on binary updates
 if (APPLE AND NOT IOS)
   if (NOT OpenSSL_DIR)
@@ -332,14 +338,8 @@ endif()
 find_package(OpenSSL REQUIRED)
 if(STATIC AND NOT IOS)
   if(UNIX)
-    set(OPENSSL_LIBRARIES "${OPENSSL_LIBRARIES};${CMAKE_DL_LIBS}")
+    set(OPENSSL_LIBRARIES "${OPENSSL_LIBRARIES};${CMAKE_DL_LIBS};${CMAKE_THREAD_LIBS_INIT}")
   endif()
-endif()
-
-if (UNIX AND NOT APPLE)
-  # Note that at the time of this writing the -Wstrict-prototypes flag added below will make this fail
-  set(THREADS_PREFER_PTHREAD_FLAG ON)
-  find_package(Threads)
 endif()
 
 add_definitions(-DAUTO_INITIALIZE_EASYLOGGINGPP)

